test/unit/orchestration_test.rb

```diff
@@ -51,6 +51,49 @@ def setTest; true; end
     end
   end
 
+  test "test host compensates queue if active record not saved" do
+    class Host::Test < Host::Base
+      include Orchestration
+
+      attr_accessor :test_flag, :progress_report_id
+
+      after_validation :queue_test
+
+      def lookup_value_match
+        "fqdn=zzz"
+      end
+
+      def queue_test
+        queue.create(
+          :name => 'test action',
+          :priority => 1,
+          :action => [self, :setTestFlags]
+        )
+      end
+
+      def setTestFlags
+        @test_flag = 'set'
+        true
+      end
+
+      def delTestFlags
+        @test_flag = 'del'
+        true
+      end
+    end
+
+    h = Host::Test.new(:name => 'test1')
+    h.stubs(:skip_orchestration?).returns(false)
+    h.stubs(:update_cache).returns(true)
+    h.stubs(:_create_record).raises(ActiveRecord::InvalidForeignKey, 'Fake foreign key exception')
+
+    assert_raise ActiveRecord::InvalidForeignKey do
+      h.save!
+    end
+
+    assert_equal 'del', h.test_flag
```

Comment:
Just use mocha expectations?  `h.expects(:setTestFlags).returns(true)` and for delTestFlags.  It'll raise at the end of the test if they weren't called, or were called unexpectedly.